prov/efa/src/efa_fork_support.c

```diff
@@ -0,0 +1,267 @@
+ * COPYING in the main directory of this source tree, or the
+ * OpenIB.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "efa.h"
+
+enum efa_fork_support_status g_efa_fork_status = EFA_FORK_SUPPORT_OFF;
+
+/**
+ * @brief initialize g_efa_fork_status based on environment and system statsu
+ *
+ */
+void efa_fork_support_request_initialize()
+{
+	g_efa_fork_status = EFA_FORK_SUPPORT_OFF;
+
+#if HAVE_IBV_IS_FORK_INITIALIZED == 1
+	if (ibv_is_fork_initialized() == IBV_FORK_UNNEEDED) {
+		g_efa_fork_status = EFA_FORK_SUPPORT_UNNEEDED;
+		return;
+	}
+#endif
+	int fork_support_requested = 0;
+
+	fi_param_get_bool(&rxr_prov, "fork_safe", &fork_support_requested);
+
+	/*
+	 * Check if any environment variables which would trigger
+	 * libibverbs' fork support are set. These variables are
+	 * defined by ibv_fork_init(3).
+	 */
+	if (fork_support_requested || getenv("RDMAV_FORK_SAFE") || getenv("IBV_FORK_SAFE"))
+		g_efa_fork_status = EFA_FORK_SUPPORT_ON;
+}
+
+/* @brief Check if rdma-core fork support is enabled and prevent fork
+ * support from being enabled later.
+ *
+ * Register a temporary buffer and call ibv_fork_init() to determine
+ * if fork support is enabled. Registering a buffer prevents future
+ * calls to ibv_fork_init() from completing successfully.
+ *
+ * This relies on internal behavior in rdma-core and is a temporary workaround.
+ *
+ * @param domain_fid domain fid so we can register memory
+ * @return 1 if fork support is enabled, 0 otherwise
+ */
+static int efa_fork_support_is_enabled(struct fid_domain *domain_fid)
+{
+	struct fid_mr *mr;
+	char *buf;
+	int ret;
+	long page_size;
+
+	/* If ibv_is_fork_initialized is availble, check if the function
+	 * can exit early.
+	 */
+#if HAVE_IBV_IS_FORK_INITIALIZED == 1
+	enum ibv_fork_status fork_status = ibv_is_fork_initialized();
+
+	/* If fork support is enabled or unneeded, return. */
```

Comment:
Not related to this refactor, but according to the code inspection, `fork_status` should not be `IBV_FORK_UNNEEDED` here because we will skip this function call for that case. So the only valid value for `fork_status` here is `IBV_FORK_ENABLED` right? Should we have a assert here?

Reply:
Yes. Though it need to be in a separate commit.

I actually think we should drop the `if (fork_status != IBV_FORK_DISABLED)`